things/postgres/groups.go

```diff
@@ -0,0 +1,631 @@
+	return toGroup(dbu), nil
+}
+
+func (gr groupRepository) RetrieveByName(ctx context.Context, name string) (groups.Group, error) {
+	q := `SELECT id, name, description, metadata FROM thing_groups WHERE name = $1`
+
+	dbu := dbGroup{
+		Name: name,
+	}
+
+	if err := gr.db.QueryRowxContext(ctx, q, name).StructScan(&dbu); err != nil {
+		if err == sql.ErrNoRows {
+			return groups.Group{}, errors.Wrap(groups.ErrNotFound, err)
+
+		}
+		return groups.Group{}, errors.Wrap(errRetrieveDB, err)
+	}
+
+	group := toGroup(dbu)
+	return group, nil
+}
+
+func (gr groupRepository) RetrieveAllParents(ctx context.Context, groupID string, offset, limit uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	if groupID == "" {
+		return groups.GroupPage{}, nil
+	}
+
+	_, mq, err := getGroupsMetadataQuery("subordinates", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("WHERE %s", mq)
+	}
+	sq := `WITH RECURSIVE subordinates (id, owner_id, parent_id, name, description, metadata, level, path) AS (
+		SELECT id, owner_id, parent_id, name, description, metadata, 0, ARRAY[thing_groups.name]
+		FROM thing_groups
+		WHERE id = :id
+		UNION
+			SELECT thing_groups.id, thing_groups.owner_id, thing_groups.parent_id, thing_groups.name, thing_groups.description, thing_groups.metadata, s.level + 1, ARRAY_PREPEND(thing_groups.name, s.path )
+			FROM thing_groups 
+			INNER JOIN subordinates s ON s.parent_id = thing_groups.id
+	)`
+	q := fmt.Sprintf("%s SELECT id, owner_id, parent_id, name, description, metadata, level, ARRAY_TO_STRING(path, '/') AS path FROM subordinates %s ORDER BY id LIMIT :limit OFFSET :offset", sq, mq)
+	cq := fmt.Sprintf("%s SELECT COUNT(*) FROM subordinates %s", sq, mq)
+	dbPage, err := toDBGroupPage("", groupID, offset, limit, gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+	defer rows.Close()
+
+	items, err := processRows(rows)
+	if err != nil {
+		return groups.GroupPage{}, err
+	}
+
+	total, err := total(ctx, gr.db, cq, dbPage)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+	}
+
+	page := groups.GroupPage{
+		Groups: items,
+		PageMetadata: groups.PageMetadata{
+			Total:  total,
+			Offset: offset,
+			Limit:  limit,
+		},
+	}
+
+	return page, nil
+}
+
+func (gr groupRepository) RetrieveAllChildren(ctx context.Context, groupID string, offset, limit uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	if groupID == "" {
+		return groups.GroupPage{}, nil
+	}
+	_, mq, err := getGroupsMetadataQuery("subordinates", gm)
+	if err != nil {
+		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+	}
+	if mq != "" {
+		mq = fmt.Sprintf("WHERE %s", mq)
+	}
+
+	sq := `WITH RECURSIVE subordinates (id, owner_id, parent_id, name, description, metadata,  level, path) AS (
```

Comment:
This query is complex and it hard to read and understand. A large part of complexity comes from having `path` and `level`. If we remove it, do we still have all the necessary data to reconstruct Group genealogy? I mean, can we construct a tree or whichever data structure we want by programmatically tracking IDs of the retrieved groups?